src/BenchmarkDotNet.TestAdapter/VSTestAdapter.cs

```diff
@@ -0,0 +1,208 @@
+                var testCases = enumerator
+                    .GetTestCasesFromSourceSerialized(source)
+                    .Select(SerializationHelpers.Deserialize<TestCase>)
+                    .ToList();
+
+                // Validate that all test ids are unique
+                var idLookup = new Dictionary<Guid, string>();
+                foreach (var testCase in testCases)
+                {
+                    if (idLookup.TryGetValue(testCase.Id, out var matchingCase))
+                        throw new Exception($"Encountered Duplicate Test ID: '{testCase.DisplayName}' and '{matchingCase}'");
+
+                    idLookup[testCase.Id] = testCase.DisplayName;
+                }
+
+                return testCases;
+            }
+            catch (Exception ex)
+            {
+                logger.SendMessage(TestMessageLevel.Error, $"Failed to load benchmarks from source\n{ex}");
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Runs the benchmarks in the given source.
+        /// </summary>
+        /// <param name="source">The dll or exe of the benchmark project.</param>
+        /// <param name="frameworkHandle">An interface used to communicate with the VSTest host.</param>
+        /// <param name="testCases">
+        /// The specific test cases to be run if specified.
+        /// If unspecified, runs all the test cases in the source.
+        /// </param>
+        private void RunBenchmarks(string source, IFrameworkHandle frameworkHandle, IEnumerable<TestCase>? testCases = null)
+        {
+            ValidateSourceOrThrow(source);
+
+            // Create a HashSet of all the TestCase IDs to be run if specified.
+            var caseIds = testCases == null ? null : new HashSet<Guid>(testCases.Select(c => c.Id));
+
+            try
+            {
+                // Ensure that test execution is done inside the context of the source directory.
+                var executor = (BenchmarkExecutorWrapper)CreateIsolatedType(typeof(BenchmarkExecutorWrapper), source);
+                cts?.Token.Register(executor.Cancel);
+
+                executor.RunBenchmarks(source, new TestExecutionRecorderWrapper(frameworkHandle), caseIds);
+            }
+            catch (Exception ex)
+            {
+                frameworkHandle.SendMessage(TestMessageLevel.Error, $"Failed to run benchmarks in source\n{ex}");
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// This will create the given type in a child AppDomain when used in .NET Framework.
+        /// If not in the .NET Framework, it will use the current
+        /// </summary>
+        /// <param name="type">The type to create.</param>
+        /// <param name="source">The dll or exe of the benchmark project.</param>
+        /// <returns>The created object.</returns>
+        private static object CreateIsolatedType(Type type, string source)
+        {
+#if NETFRAMEWORK
```

Comment:
I think it'd make more sense to do a runtime lookup than multi-targeting. `if (RuntimeInformation.IsFullFramework)`

Reply:
I'm unable to do this as `AppDomainSetup` and related methods are not supported in `netstandard2.0`. They are supported in `net6.0` though, but I'm not sure how to get around this without the multi-targeting.

Reply:
Is the `AppDomainSetup` required? Can you just use the `CreateDomain(string)` version that's part of netstandard2.0?

Reply:
Unfortunately it's needed for .NET Framework so I can set the `ApplicationBase` for the new domain so that it will load all the assemblies from the BDN directory rather than from the VSTest directory. There is no supported cross-platform way to achieve this without cross-targeting.

See: https://github.com/dotnet/runtime/issues/28026#issuecomment-444562466

An alternative though would be that we don't support the VSTest Adapter for .NET Framework projects

Reply:
Another alternative would be to use reflection to use that API. But I'm not strong-minded about it. It's probably fine to multi-target if all the testing libraries do the same.